pkg/postgres/pgutils/retry.go

```diff
@@ -13,36 +14,42 @@ const (
 
 // Retry is used to specify how long to retry to successfully run a query with 1 return value
 // that fails with transient errors
-func Retry(fn func() error) error {
+func Retry(ctx context.Context, fn func() error) error {
 	// Shape fn to match the Retry2 below
 	fnWithReturn := func() (struct{}, error) {
 		return struct{}{}, fn()
 	}
-	_, err := Retry2(fnWithReturn)
+	_, err := Retry2(ctx, fnWithReturn)
 	return err
 }
 
 // Retry2 is used to specify how long to retry to successfully run a query with 2 return values
 // that fails with transient errors
-func Retry2[T any](fn func() (T, error)) (T, error) {
+func Retry2[T any](ctx context.Context, fn func() (T, error)) (T, error) {
 	// Shape fn to match the Retry3 below
 	fnWithReturn := func() (T, struct{}, error) {
 		val, err := fn()
 		return val, struct{}{}, err
 	}
-	val, _, err := Retry3(fnWithReturn)
+	val, _, err := Retry3(ctx, fnWithReturn)
 	return val, err
 }
 
 // RetryIfPostgres is same as Retry except that it retries iff postgres is enabled.
 // that fails with transient errors
-func RetryIfPostgres(fn func() error) error {
-	return Retry(fn)
+func RetryIfPostgres(ctx context.Context, fn func() error) error {
+	return Retry(ctx, fn)
 }
 
 // Retry3 is used to specify how long to retry to successfully run a query with 3 return values
 // that fails with transient errors
-func Retry3[T any, U any](fn func() (T, U, error)) (T, U, error) {
+func Retry3[T any, U any](ctx context.Context, fn func() (T, U, error)) (T, U, error) {
+	var ret1 T
+	var ret2 U
+	if err := ctx.Err(); err != nil {
+		return ret1, ret2, err
+	}
```

Comment:
Do we need this short circuit here? Will it not be handled in `select` later?

Reply:
I added the short circuit because I think it is the more correct behavior. That is, if the context is expired, we never try even once.